First version

pmenendz opened the pull request:
I created the gobl.tin only with the functionality of checking the VIES VAT numbers (for the moment). To allow scalability and include other databases in the future I created an interface (TinLookup) and a factory (tin_factory.go).

pmenendz committed:
VIES lookup

pmenendz committed:
Testing vies

pmenendz committed:
test for factory

pmenendz committed:
testing with gobl invoices

pmenendz committed:
Create LICENSE

pmenendz committed:
Documenting

pmenendz committed:
Removing debugging prints

samlown commented on gobltin.go:
```diff
@@ -0,0 +1,64 @@
+}
+
+// NewTinNumber creates a new TinNumber object from an envelope
+func NewTinNumber(env *gobl.Envelope) (*TinNumber, error) {
```
I think the verb here is more like "Extract"... the idea being that it should be capable of finding `tax.Identity` entries from inside the document. It may also be worth considering returning an array.

Another consideration here is that by "extracting" the tax identities, you're effectively loosing their location inside the parent structure, which would be useful for output...

samlown commented on gobltin.go:
```diff
@@ -0,0 +1,64 @@
+	"github.com/invopop/gobl/bill"
+)
+
+type TinNumber struct {
```
Rather than defining a new structure, I'd suggest using the existing `tax.Identity` model from GOBL.

samlown commented on gobltin.go:
```diff
@@ -0,0 +1,64 @@
+	}
+
+	// Check if the customer has a tax ID
+	if inv.Customer.TaxID == nil {
```
This is a bit "brittle", most likely as a user you wouldn't expect an error if there are simply no tax IDs to validate.

samlown commented on gobltin.go:
```diff
@@ -0,0 +1,64 @@
+
+}
+
+func (t *TinNumber) Lookup() (*CheckTinResponse, error) {
```
Rather than centralising the validation on a `TinNumber`, I'd suggest a model more like a `Client` that encapsulates the logic. So as a user you can do:
```
c := gobltin.New()
if err := c.Lookup(ctx, tin); err != nil {
  return err
}
```

samlown commented on tin_lookup.go:
```diff
@@ -0,0 +1,13 @@
+}
+
+type TinLookup interface {
+	LookupTin(countryCode, tinNumber string) (*CheckTinResponse, error)
```
You want to include a "context" here, which is an unusual but highly recommended convention in Go. I'd also recommend just passing in the GOBL `tax.Identity` object:

```suggestion
	LookupTin(ctx context.Context, tid *tax.Identity) (*CheckTinResponse, error)
```

samlown commented on vies.go:
```diff
@@ -0,0 +1,59 @@
+		return nil, errors.New("error marshalling JSON")
+	}
+
+	resp, err := http.Post(VIES_API_URL, "application/json", bytes.NewBuffer(jsonData))
```
I normally recommend using a library like "resty" for wrapping around HTTP requests. It makes it much easier to debug, and handles a lot of the nuances around making connections.

samlown commented on tin_factory.go:
```diff
@@ -0,0 +1,28 @@
+}
+
+// List of all EU country codes supported by VIES
+var europeanCountryCodes = []string{
```
I'd suggest trying to leverage as much as possible from GOBL for the different types. Instead of an array of `string`, I'd consider `[]l10n.Code` directly. You could also use the constants pre-defined in the GOBL `l10n` package.

samlown reviewed (changes requested):
This is looking really good! Even more so for a first attempt, nicely done. I've made a few suggestions on how to approach the structure of the app that I think are worth considering.

I think we may also need to enable linting, as there are few things that would be picked up, like comments on methods.

pmenendz committed:
adding cmd and using tax.Identity

pmenendz committed:
Linting and changing structure

pmenendz commented:
I included @samlown comments, and I changed the structure especially the gobltin.go, now it returns an array that depends if the user wants to validate the supplier, consumer or both. I created constants for this, I know that there are elements already in the gobl (org.Party), but I only needed constants in order to validate one or the other, so I wasn't sure how to integrate it with the org.Party variables. 

The main changes:

- Use resty library instead of http
- Use type l10n.CountryCode for the country codes
- Now the user can choose to validate the customer (by default), supplier or both.
- Created 2 types of errors: InvalidId and Network. The first one doesn't return the user an error but a false in the response and a message, while the second one returns an error.
- Using type tax.Identity
- Linted using golangci-lint

I also included a cmd interaction. 

Thanks for the comments and feedback.

pmenendz committed:
Making it more general and simpler

pmenendz committed:
Moving context to main function

pmenendz committed:
test issue

samlown commented on .github/workflows/lint.yaml:
```diff
@@ -0,0 +1,27 @@
+      - name: Set up Go
+        uses: actions/setup-go@v4
+        with:
+          go-version: "1.22.3"
```
I'd suggest using the go.mod to extract the version. Example here: https://github.com/invopop/transform/blob/main/.github/workflows/lint.yaml#L15-L22

samlown commented on gobltin.go:
```diff
@@ -0,0 +1,37 @@
+//
+// The user can choose to validate the customer, the supplier, or both, included as an argument.
+// If the party type is not provided, the function will validate only the customer.
+func LookupTin(ctx context.Context, party *org.Party) (bool, error) {
```
I'd remove the `Tin` part of the name as we already know its a "TIN" from the package name. Also, Go convention suggests that abbreviations should be capitalised, so the correct method name would be `LookupTIN`. (I'd go with just `Lookup` though).

You might always want to consider making this method more generic. You could accept `any` instead of `org.Party` then try to work out what you can check based on the type.

pmenendz replied:
At first, I had done it for an invoice as input, but in that case there should also be another argument that is if you want to validate the customer, the supplier or both. What is more, if it is an invoice and you are validating both, the function will return an array specifying which one is the customer and the supplier. This is what I had before and I believed it was added complexity when the user can just do two calls to the function when they need the customer and the supplier. Let me know what you think.

Here it is how I did it before: https://github.com/invopop/gobl.tin/pull/1/commits/c43c4065af4e93141bd1fa06c34ed321b46d32dd#diff-43314b5b16ef2102e9657b1f734c35a18a913605c547b6a6f6ed60381e758d5a:~:text=only%20the%20customer.-,func%20LookupTin(env%20*gobl.Envelope%2C%20args%20...,%7D,-62%20changes%3A%2043

pmenendz replied:
I agree in any case that it would be useful to have an any for the case the user just wants to check the id and not the org.Party

samlown commented on tin_factory.go:
```diff
@@ -0,0 +1,31 @@
+import "github.com/invopop/gobl/l10n"
+
+// GetTinLookup returns the TinLookup for a given country code
+func GetTinLookup(countryCode l10n.CountryCode) TinLookup {
```
Again, you want something like `GetTINLookup`.

samlown commented on tin_factory.go:
```diff
@@ -0,0 +1,31 @@
+// List of all EU country codes supported by VIES
+// XI is used for Northern Ireland when needs to be distinguished from UK
+var europeanCountryCodes = []l10n.CountryCode{
+	"AT", "BE", "BG", "CY", "CZ", "DE", "DK", "EE", "EL", "ES", "FI",
```
If you use constants here instead of text, you get validation for free.

samlown commented on vies.go:
```diff
@@ -0,0 +1,68 @@
+package gobltin
```
I'd be tempted to consider putting these implementations inside their own internal packages rather than in the main namespace. That'll scale better as potentially more implementations are added.

samlown commented on vies.go:
```diff
@@ -0,0 +1,68 @@
+}
+
+// CommonResponse is the response body for the VIES API
+type CommonResponse struct {
```
This is a generic name for a specific namespace in the public package. If this is in an independent package it'll scale a bit better.

samlown reviewed (dismissed):
Looking Good! I'd consider some of the naming and packages here. Might even be worth just renaming the package `tin` instead of `gobl.tin`. I'd also renaming `Tin` to `TIN` to be consistent with Go naming.

pmenendz committed:
Resolving conversations

pmenendz committed:
Linting

pmenendz committed:
Making it more general

pmenendz committed:
adding cache

pmenendz committed:
Updating Readme

pmenendz committed:
Returning only error

pmenendz committed:
Country code Greece and Northern Ireland

pmenendz commented:
What it has changed from the last review:
- Now there is an API package and one package for each of the apis (now only VIES)
- I also added a cache for caching already validated TINs
- Changed names to make it easier to read
- Made it more general (3 input options: invoice, party and taxid)
- List of country codes with constants instead of string
- Now it only returns an error

samlown reviewed (approved):
Cool! Looks good!
